### PR TITLE
Second Sprint

### PR DESCRIPTION
**Pull request by WojTechM**
Games between 2 players are happening - communication between client and server works
Boards players use are (for now) predefined in code.

### CODE REVIEW

File: server/src/main/java/com/sarny/spocone/server/game/Round.java [hunk 1]
```diff
@@ -1,7 +1,85 @@
...
+
+    ShotsSummary getOpponentsShots(int playerID) {
+        int oppositePlayerID = oppositePlayerID(playerID);
+        List<ShotResult> shotResults = playersShots.get(oppositePlayerID);
+        return new ShotsSummary(shotResults, checkIfLastShotIsWin(shotResults));
+    }
+
+    Integer lastSunkField(int playerID) {
+        return playersShots.get(playerID).stream()
+                .filter(r -> r.getShotOutcome() == ShotOutcome.SUNK || r.getShotOutcome() == ShotOutcome.WIN)
```
Comment (WojTechM) [inline server/src/main/java/com/sarny/spocone/server/game/Round.java]: change r to something descriptive

File: server/src/main/java/com/sarny/spocone/server/game/ShipNeighbouringFieldsGenerator.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+class ShipNeighbouringFieldsGenerator {
+
+    private static final int COLUMNS = 10;
+
+    Set<Integer> generateNeighbours(Ship ship) {
+        Set<Integer> neighbours = new HashSet<>();
+        List<Integer> occupiedFields = getOccupiedFields(ship);
+
+        for (Integer field : occupiedFields) {
+            if (field != (((field / COLUMNS) * COLUMNS) + COLUMNS - 1)) {
```
Comment (WojTechM) [inline server/src/main/java/com/sarny/spocone/server/game/ShipNeighbouringFieldsGenerator.java]: extract descriptive method to if conditions statements.

File: server/src/main/java/com/sarny/spocone/server/game/ShipNeighbouringFieldsGenerator.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                generateNeighboursFromLeftSideOfAField(field, neighbours, ship.hit);
+            }
+            generateNeighboursAboveAndUnderAField(field, neighbours, ship.hit);
+
+        }
+        return neighbours;
+    }
+
+    private void generateNeighboursAboveAndUnderAField(int field, Set<Integer> neighbours, List<Integer> ship) {
+        addNeighbourIfNotAShip(field - COLUMNS, neighbours, ship);
```
Comment (WojTechM) [inline server/src/main/java/com/sarny/spocone/server/game/ShipNeighbouringFieldsGenerator.java:35]: addNeighbourIfIsNotAShip

File: server/src/main/java/com/sarny/spocone/server/game/ShipPlacementValidator.java [hunk 4]
```diff
@@ -1,8 +1,125 @@
...
 /**
  * @author Wojciech Makiela
  */
 class ShipPlacementValidator {
 
+    private final Board board;
+    private final ShipNeighbouringFieldsGenerator generator;
+    private final Map<Integer, Integer> shipsOfLengthToPlace;
+
+    private ShipPlacementValidator(HashMap<Integer, Integer> shipsOfLengthToPlace, Board board, ShipNeighbouringFieldsGenerator generator) {
```
Comment (WojTechM) [inline server/src/main/java/com/sarny/spocone/server/game/ShipPlacementValidator.java]: Change HashMap to Map in constructor.

File: server/src/main/java/com/sarny/spocone/server/game/ShipPlacementValidator.java [hunk 5]
```diff
@@ -1,8 +1,125 @@
...
+    }
+
+    void placedNewShip(Ship ship) {
+        int previousRemainingShips = shipsOfLengthToPlace.get(ship.length());
+        shipsOfLengthToPlace.put(ship.length(), previousRemainingShips - 1);
+    }
+
+    static class Builder implements WithShipsOfLength4, WithShipsOfLength3, WithShipsOfLength2, WithShipsOfLength1, WithGuaranteedMissGenerator, BuildShipPlacementValidator {
+
+        private int length4 = 1;
```
Comment (WojTechM) [inline server/src/main/java/com/sarny/spocone/server/game/ShipPlacementValidator.java]: length3 caused i was confused for few seconds. Change to lengthOfShip3 -> more readable

**Review by atrzewik — CHANGES_REQUESTED** [review]
Nice Interface segregation. Some empty classes - to implement in the future.
Code is quite readable, nice test coverage, but number of tests - 112 is low.
Fix commented issues.

**Pull request merged**